File: 02-linked-lists/20250726-beginner/go/solution.go
package main

import "fmt"

// Node represents a single node in the linked list
type Node struct {
	Data int
	Next *Node
}

// LinkedList represents the linked list structure
type LinkedList struct {
	Head *Node
}

// NewLinkedList creates a new empty linked list
func NewLinkedList() *LinkedList {
	// TODO: Implement constructor
	newList := new(LinkedList)
	return newList
}

// Append adds a new node with the given data at the end of the list
func (ll *LinkedList) Append(data int) {
	newNode := &Node{Data: data, Next: nil}
	if ll.Head == nil {
		ll.Head = newNode
		return
	}
	current := ll.Head
	for current.Next != nil {
		current = current.Next
	}
	current.Next = newNode
}

// ToSlice returns all elements in the list as a slice
func (ll *LinkedList) ToSlice() []int {
	// TODO: Implement conversion to slice
	s := make([]int, 0)
	if ll.Head == nil {
		fmt.Println("early slice exit")
		return s
	}
	current := ll.Head
	s = append(s, current.Data)
	for current.Next != nil {
		current = current.Next
		s = append(s, current.Data)
	}
	return s
}

// Reverse reverses the linked list in-place
func (ll *LinkedList) Reverse() {
	// TODO: Implement in-place reversal of the linked list
	if ll.Head == nil {
		return
	}
	var previous *Node = nil
	var next *Node
	current := ll.Head
	for current != nil {
		next = current.Next
		current.Next = previous
		previous = current
		current = next

	}
	ll.Head = previous

}

// Size returns the number of elements in the list
func (ll *LinkedList) Size() int {
	if ll.Head == nil {
		return 0
	}
	count := 1
	current := ll.Head
	for current.Next != nil {
		count++
		current = current.Next
	}
	return count
}

File: 02-linked-lists/20250726-beginner/ts/solution.ts
// Node class represents a single node in the linked list
class Node {
    data: number;
    next: Node | null;

    constructor(data: number) {
        this.data = data;
        this.next = null;
    }
}

// LinkedList class represents the linked list structure
export class LinkedList {
    private head: Node | null;

    constructor(head?: Node) {
        this.head = head ? head : null;
    }

    // Append adds a new node with the given data at the end of the list
    append(data: number): void {
        const newNode = new Node(data)
        if (this.head === null) {
            this.head = newNode
            return
        }
        let current = this.head
        while (current != null) {
            console.log("append loop  start")
            if (current.next != null) {
                current = current.next
            } else {
                current.next = newNode
                return
            }
        }
    }

    // toArray returns all elements in the list as an array
    toArray(): number[] {
        if (this.head === null) {
            return []
        }
        let current: Node | null = this.head
        let out = []
        while (current !== null) {
            console.log("array loop  start")
            out.push(current.data)
            current = current.next
        }
        console.log(out)
        return out
    }

    // reverse reverses the linked list in-place
    reverse(): void {
        // TODO: Implement in-place reversal of the linked list
        // This should modify the existing list structure, not create a new one
        if (this.head === null) {
            return
        }
        let next: Node | null = null
        let current: Node | null = this.head
        let previous: Node | null = null
        while (current !== null) {
            console.log("reverse loop start")
            next = current.next
            current.next = previous
            previous = current
            current = next
            
        }
        this.head = previous
        return
    }

    // size returns the number of elements in the list
    size(): number {
        let count = 0
        let current: Node | null = this.head
        while (current !== null) {
            count ++
            current = current.next
            
        }
        return  count
    }
}
